### PR TITLE
add section on overlap checks

### PR DESCRIPTION
**Pull request by jdonszelmann**
Added a section on overlap checks after talking with @lcnr

### CODE REVIEW

**Review by lcnr — COMMENTED** [review]
please move the chapter `Type checking` and rename it to the more general "Coherence checking". Talking about overlap checking outside of coherence feels like it may lead to the wrong impression.

File: src/traits/overlap.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Overlap checks
+
+As part of checking items (specifically: structs, enums, traits, unions),
```
[the remark this answers is not in the document]
Comment (jdonszelmann) [reply src/traits/overlap.md:3]: I'll see what I can do :)

**Comment by jdonszelmann** [comment]
I didn't know too much more to say specifically on the orphan check, maybe that's not needed as much either. However, I did put it in the broader context of coherence checking (and it's now moved under type checking)

File: src/coherence.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+This check tries to find a negative trait implementation. 
+For example:
+
+```rust
+struct MyCustomBox<T: ?Sized>(Box<T>)
+
+// both in your own crate
+impl From<&str> for MyCustomBox<dyn Error> {}
+impl<E> From<E> for MyCustomBox<dyn Error> where E: Error {}
```
Comment (lcnr) [inline src/coherence.md:61]:
```suggestion
struct MyCustomErrorType;

// both in your own crate
impl From<&str> for MyCustomErrorType {}
impl<E> From<E> for MyCustomErrorType where E: Error {}
```

File: src/coherence.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+```
+
+In this example, we'd get:
+`MyCustomBox<dyn Error>: From<&str>` and `MyCustomBox<dyn Error>: From<?E>`, giving `?E = &str`.
```
Comment (lcnr) [inline src/coherence.md:65]:
```suggestion
`MyCustomErrorType: From<&str>` and `MyCustomErrorType: From<?E>`, giving `?E = &str`.
```

File: src/coherence.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+And thus, these two implementations would overlap.
+However, libstd provides `&str: !Error`, and therefore guarantees that there 
+will never be a positive implementation of `&str: Error`, and thus there is no overlap.
+
+Note that for this kind of negative impl check, we must have explicit negative implementations provided.
+This is not currently stable.
+
+[`impl_intersection_has_negative_obligation`]: https://doc.rust-lang.org/beta/nightly-rustc/rustc_trait_selection/traits/coherence/fn.impl_intersection_has_impossible_obligation.htmlhttps://doc.rust-lang.org/beta/nightly-rustc/rustc_trait_selection/traits/coherence/fn.impl_intersection_has_negative_obligation.html
```
Comment (lcnr) [inline src/coherence.md:74]:
```suggestion
[`impl_intersection_has_negative_obligation`]: https://doc.rust-lang.org/beta/nightly-rustc/rustc_trait_selection/traits/coherence/fn.impl_intersection_has_impossible_obligation.html
```

**Review by lcnr — APPROVED** [review]

**Pull request merged**